keps/sig-api-machinery/20190325-unions.md

```diff
@@ -214,24 +238,74 @@ below.
-    to that value. Otherwise,
-  - compare the fields set before and after. If there is exactly one field
-    added, set the discriminator (if present) to that value, and remove all
-    other fields. if more than one field has been added, leave the process so
-    that validation will fail.
+- If there is more than one field added in the union (compared to the old
+  object), fail with an error,
+- If the discriminator has changed to a non-nil/non-empty value, and there is
+  one new field set but they don't match, return an error,
+- If the discriminator has changed to a non-nil/non-empty value, clear all the
+  fields but the discriminated field,
+- If only one field is left, update a deducible discriminator to that value,
+- if more than one field is left, fail with an error.
+
+Example updates with valid content in the new object:
+| Scenario | New object                       | Existing object                  | Result                           |
+|----------|----------------------------------|----------------------------------|----------------------------------|
+| A1       | empty                            | empty                            | empty                            |
+| A2       | empty                            | `a: {...}`                       | empty                            |
+| A3       | empty                            | `discriminator: A`<br>`a: {...}` | empty                            |
+| A4       | `a: {...}`                       | empty                            | `discriminator: A`<br>`a: {...}` |
+| A5       | `a: {...}`                       | `a: {...}`                       | `discriminator: A`<br>`a: {...}` |
+| A6       | `a: {...}`                       | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` |
+| A7       | `b: {...}`                       | empty                            | `discriminator: B`<br>`b: {...}` |
+| A8       | `b: {...}`                       | `a: {...}`                       | `discriminator: B`<br>`b: {...}` |
+| A9       | `b: {...}`                       | `discriminator: A`<br>`a: {...}` | `discriminator: B`<br>`b: {...}` |
+| A10      | `discriminator: A`<br>`a: {...}` | empty                            | `discriminator: A`<br>`a: {...}` |
+| A11      | `discriminator: A`<br>`a: {...}` | `a: {...}`                       | `discriminator: A`<br>`a: {...}` |
+| A12      | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` |
+| A13      | `discriminator: B`<br>`b: {...}` | empty                            | `discriminator: B`<br>`b: {...}` |
+| A14      | `discriminator: B`<br>`b: {...}` | `a: {...}`                       | `discriminator: B`<br>`b: {...}` |
+| A15      | `discriminator: B`<br>`b: {...}` | `discriminator: A`<br>`a: {...}` | `discriminator: B`<br>`b: {...}` |
+
+Example updates with invalid content in the new object:
```

Comment:
There's two ways of getting to the things in this table: via PUT or via PATCH. via PUT I agree with failing requests, clients should not explicitly ask for invalid states, and most clients drop fields they don't understand anyway, which would produce an entry in the table above.

However clients that want to PATCH adding just a field they understand might produce a state in this table. I think such requests are not ambiguous, and I'm not thrilled about having them lumped in here.

Reply:
I see two ways we could handle that:
1. We could parse each type of patch (merge patch, json merge patch and strategic merge patch) and detect the intent from each patch to know if the intent makes sense and then try to do the right thing.
2. We could try to deduce the intent of the patch based on pre-patch and post-patch (only when a patch is sent, not for updates) and apply the deduce logic in that case.

I suspect that 1 is going be both hard to implement and maintain, so I'm much more in favor of 2.
This would simply mean that we have two separate logic, one for update (no deduce) and one for patch (deduce). I can do that if you guys think that's a reasonable strategy.

Reply:
Yes, 2 was what I had in mind.

@liggitt points out that this will accept some PATCHes that really ought to be rejected.

I really object to requiring good citizens to do extra work (set the discriminator in patches) just to make sure we stick it to the bad citizens, but I have too many other PRs to review right now to make extra sure that on balance this helps more than it hurts.